src/libpgagroal/ev.c

```diff
@@ -0,0 +1,2037 @@
+   ret = periodic_stop(ev, target);
+   list_delete(w, &ev->phead.next, target, ret);
+   return ret;
+}
+
+static int
+setup_ops(struct ev_loop* ev)
+{
+   int ret = EV_OK;
+   bool mtt = ev->ctx.multithreading;
+   struct configuration* config = (struct configuration*)shmem;
+
+   is_running = mtt ? __is_running_atomic : __is_running;
+   set_running = mtt ? __set_running_atomic: __set_running;
+   loop_break = mtt ? __break_atomic: __break;
+
+   if (!strcmp(config->ev_backend, "io_uring"))
+   {
+#if HAVE_URING
+      loop_init = __io_uring_init;
+      loop_fork = __io_uring_fork;
+      loop_destroy = __io_uring_destroy;
+      loop_start = __io_uring_loop;
+      io_start = __io_uring_io_start;
+      io_stop = __io_uring_io_stop;
+      periodic_init = __io_uring_periodic_init;
+      periodic_start = __io_uring_periodic_start;
+      periodic_stop = __io_uring_periodic_stop;
+      signal_start = __io_uring_signal_start;
+      signal_stop = __io_uring_signal_stop;
+#endif
+   }
+   if (!strcmp(config->ev_backend, "epoll"))
+   {
+#if HAVE_EPOLL
+      loop_init = __epoll_init;
+      loop_fork = __epoll_fork;
+      loop_destroy = __epoll_destroy;
+      loop_start = __epoll_loop;
+      io_start = __epoll_io_start;
+      io_stop = __epoll_io_stop;
+      periodic_init = __epoll_periodic_init;
+      periodic_start = __epoll_periodic_start;
+      periodic_stop = __epoll_periodic_stop;
+      signal_start = __epoll_signal_start;
+      signal_stop = __epoll_signal_stop;
+#endif
+   }
+   else if (!strcmp(config->ev_backend, "kqueue"))
+   {
+#if HAVE_KQUEUE
+      loop_init = __kqueue_init;
+      loop_fork = __kqueue_fork;
+      loop_destroy = __kqueue_destroy;
+      loop_start = __kqueue_loop;
+      io_start = __kqueue_io_start;
+      io_stop = __kqueue_io_stop;
+      periodic_init = __kqueue_periodic_init;
+      periodic_start = __kqueue_periodic_start;
+      periodic_stop = __kqueue_periodic_stop;
+      signal_start = __kqueue_signal_start;
+      signal_stop = __kqueue_signal_stop;
+#endif
+   }
+
+   return ret;
+}
+
+/*
+ * TODO: move this to libpgagroal/configuration.c and allow configuration
+ */
+static int
+setup_context(struct ev_context* ctx)
+{
+   struct configuration* config = &((struct main_configuration*)shmem)->common;
+   /* ordered from highest to lowest priority */
+   char* backends[] = {
+#if HAVE_URING
+      "io_uring",
+#endif
+#if HAVE_EPOLL
+      "epoll",
+#endif
+#if HAVE_KQUEUE
+      "kqueue",
+#endif
+   };
+   char log[] = (
+#if HAVE_URING
+      "io_uring, "
+#endif
+#if HAVE_EPOLL
+      "epoll, "
+#endif
+#if HAVE_KQUEUE
+      "kqueue, "
+#endif
+      );
+
+   if (sizeof(backends) == 0)
+   {
+      pgagroal_log_fatal("no ev_backend supported");
+      exit(1);
+   }
+
+   log[strlen(log) - 2] = '\0';
+   pgagroal_log_debug("Available ev backends: %s", log);
+
+   if (!strnlen(config->ev_backend, MISC_LENGTH))
+   {
+      pgagroal_log_warn("ev_backend not set in configuration file");
```

Comment:
This is log_debug

Reply:
> This is log_debug

I would prefer to have it as a warning, or info, since it is a message someone has to "catch" because it leads to a possible misconfigured system.